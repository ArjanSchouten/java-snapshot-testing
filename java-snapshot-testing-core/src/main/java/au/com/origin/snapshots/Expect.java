package au.com.origin.snapshots;

import au.com.origin.snapshots.comparators.SnapshotComparator;
import au.com.origin.snapshots.reporters.SnapshotReporter;
import au.com.origin.snapshots.serializers.SnapshotSerializer;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;

@RequiredArgsConstructor
public class Expect {
  private final SnapshotVerifier snapshotVerifier;
  private final Method testMethod;

  private SnapshotSerializer snapshotSerializer;
  private SnapshotComparator snapshotComparator;
  private List<SnapshotReporter> snapshotReporters;
  private String scenario;

  public static Expect of(SnapshotVerifier snapshotVerifier, Method method) {
    return new Expect(snapshotVerifier, method);
  }

  /**
   * Make an assertion on the given input parameters against what already exists
   *
   * @param firstObject first snapshot object
   * @param objects     other snapshot objects
   */
  public void toMatchSnapshot(Object firstObject, Object... objects) {
    Snapshot snapshot = snapshotVerifier.expectCondition(testMethod, firstObject, objects);
    if (snapshotSerializer != null) {
      snapshot.setSnapshotSerializer(snapshotSerializer);
    }
    if (snapshotComparator != null) {
      snapshot.setSnapshotComparator(snapshotComparator);
    }
    if (snapshotReporters != null) {
      snapshot.setSnapshotReporters(snapshotReporters);
    }
    if (scenario != null) {
      snapshot.setScenario(scenario);
    }
    snapshot.toMatchSnapshot();
  }


  /**
   * Normally a snapshot can be applied only once to a test method.
   * <p>
   * For Parameterized tests where the same method is executed multiple times you can supply
   * the scenario() to overcome this restriction.  Ensure each scenario is unique.
   *
   * @param scenario - unique scenario description
   * @return Snapshot
   */
  public Expect scenario(String scenario) {
    this.scenario = scenario;
    return this;
  }

  /**
   * Apply a custom serializer for this snapshot
   *
   * @param serializer your custom serializer
   * @return Snapshot
   */
  public Expect serializer(SnapshotSerializer serializer) {
    this.snapshotSerializer = serializer;
    return this;
  }

  /**
   * Apply a custom serializer for this snapshot
   *
   * @param name - the {name} attribute serializer.{name} from snapshot.properties
   * @return Snapshot
   */
  public Expect serializer(String name) {
    this.snapshotSerializer = SnapshotProperties.getInstance("serializer." + name);
    return this;
  }

  /**
   * Apply a custom comparator for this snapshot
   *
   * @param comparator your custom comparator
   * @return Snapshot
   */
  public Expect comparator(SnapshotComparator comparator) {
    this.snapshotComparator = comparator;
    return this;
  }

  /**
   * Apply a list of custom reporters for this snapshot
   * This will replace the default reporters defined in the config
   *
   * @param reporters your custom reporters
   * @return Snapshot
   */
  public Expect reporters(SnapshotReporter... reporters) {
    this.snapshotReporters = Arrays.asList(reporters);
    return this;
  }

  /**
   * Apply a custom serializer for this snapshot.
   *
   * @param serializer your custom serializer
   * @return this
   * @see au.com.origin.snapshots.serializers.SnapshotSerializer
   * <p>
   * Example implementations
   * @see au.com.origin.snapshots.serializers.ToStringSnapshotSerializer
   * @see au.com.origin.snapshots.serializers.Base64SnapshotSerializer
   */
  @SneakyThrows
  public Expect serializer(Class<? extends SnapshotSerializer> serializer) {
    this.snapshotSerializer = serializer.getConstructor().newInstance();
    return this;
  }

}
