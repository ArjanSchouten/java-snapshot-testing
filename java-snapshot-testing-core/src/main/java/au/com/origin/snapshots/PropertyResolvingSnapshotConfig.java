package au.com.origin.snapshots;

import au.com.origin.snapshots.comparators.SnapshotComparator;
import au.com.origin.snapshots.reporters.SnapshotReporter;
import au.com.origin.snapshots.serializers.SnapshotSerializer;
import org.apache.commons.lang3.StringUtils;

import java.util.List;

public class PropertyResolvingSnapshotConfig implements SnapshotConfig {

  @Override
  public String getOutputDir() {
    return SnapshotProperties.getOrThrow("output-dir");
  }

  @Override
  public String getSnapshotDir() {
    return SnapshotProperties.getOrThrow("snapshot-dir");
  }

  @Override
  public SnapshotSerializer getSerializer() {
    return SnapshotProperties.getInstance("serializer");
  }

  @Override
  public SnapshotComparator getComparator() {
    return SnapshotProperties.getInstance("comparator");
  }

  @Override
  public List<SnapshotReporter> getReporters() {
    return SnapshotProperties.getInstances("reporters");
  }

  @Override
  public boolean isCI() {
    String envVariable = SnapshotProperties.getOrThrow("ci-env-var");
    return StringUtils.isNotEmpty(System.getenv(envVariable));
  }
}
