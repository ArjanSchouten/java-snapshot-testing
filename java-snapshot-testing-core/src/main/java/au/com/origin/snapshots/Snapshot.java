package au.com.origin.snapshots;

import org.apache.commons.lang3.StringUtils;
import org.assertj.core.util.diff.DiffUtils;
import org.assertj.core.util.diff.Patch;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.Collection;
import java.util.Set;
import java.util.function.Function;

class Snapshot {

    private final SnapshotConfig snapshotConfig;
    private final SnapshotFile snapshotFile;
    private final Class testClass;
    private final Method testMethod;
    private final Function<Object, String> jsonFunction;
    private final Object[] current;

    private String scenario = null;

    Snapshot(
            SnapshotConfig snapshotConfig,
            SnapshotFile snapshotFile,
            Class testClass,
            Method testMethod,
            Function<Object, String> jsonFunction,
            Object... current) {
        this.snapshotConfig = snapshotConfig;
        this.current = current;
        this.snapshotFile = snapshotFile;
        this.testClass = testClass;
        this.testMethod = testMethod;
        this.jsonFunction = jsonFunction;
    }

    /**
     * Normally a snapshot can be applied only once to a test method.
     *
     * For Parameterized tests where the same method is executed multiple times you can supply
     * the scenario() to overcome this restriction.  Ensure each scenario is unique.
     *
     * @param scenario - unique scenario description
     */
    public Snapshot scenario(String scenario) {
        this.scenario = scenario;
        return this;
    }

    public void toMatchSnapshot() {

        Set<String> rawSnapshots = snapshotFile.getRawSnapshots();

        String rawSnapshot = getRawSnapshot(rawSnapshots);

        String currentObject = takeSnapshot();

        // Match Snapshot
        if (rawSnapshot != null && !snapshotConfig.shouldUpdateSnapshot()) {
            if (!rawSnapshot.trim().equals(currentObject.trim())) {
                throw generateDiffError(rawSnapshot, currentObject);
            }
        }
        // Create New Snapshot
        else {
            snapshotFile.push(currentObject);
        }
    }

    private SnapshotMatchException generateDiffError(String rawSnapshot, String currentObject) {
        // compute the patch: this is the diffutils part
        Patch<String> patch =
            DiffUtils.diff(
                Arrays.asList(rawSnapshot.trim().split("\n")),
                Arrays.asList(currentObject.trim().split("\n")));
        String error =
            "Error on: \n"
                + currentObject.trim()
                + "\n\n"
                + patch
                .getDeltas()
                .stream()
                .map(delta -> delta.toString() + "\n")
                .reduce(String::concat)
                .get();
        return new SnapshotMatchException(error);
    }

    private String getRawSnapshot(Collection<String> rawSnapshots) {
        for (String rawSnapshot : rawSnapshots) {

            if (rawSnapshot.contains(getSnapshotName())) {
                return rawSnapshot;
            }
        }
        return null;
    }

    private String takeSnapshot() {
        return getSnapshotName() + jsonFunction.apply(current);
    }

    String getSnapshotName() {
        String scenarioFormat = StringUtils.isBlank(scenario) ? "" : "[" + scenario + "]";
        return testClass.getName() + "." + testMethod.getName() + scenarioFormat + "=";
    }
}
