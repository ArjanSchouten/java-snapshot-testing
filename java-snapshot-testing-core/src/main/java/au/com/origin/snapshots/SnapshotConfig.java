package au.com.origin.snapshots;

import au.com.origin.snapshots.serializers.SnapshotSerializer;
import au.com.origin.snapshots.serializers.ToStringSnapshotSerializer;

import java.lang.reflect.Method;
import java.util.Optional;

/**
 *  Snapshot Configuration
 *  ----------------------
 *
 *  Implement this interface when integrating `java-snapshot-testing` with a custom testing library
 *
 *  For frameworks - consider extending the framework implementation instead
 */
public interface SnapshotConfig {
    String JVM_UPDATE_SNAPSHOTS_PARAMETER = "updateSnapshot";

    /**
     * The base directory where files get written (excluding package directories)
     * default: "src/test/java"
     *
     * You might want to override if you have tests under "src/test/integration" for example
     *
     * @deprecated Use getOutputDir() instead
     * @return test src folder
     */
    @Deprecated
    default String getTestDir() {
        return getOutputDir();
    }

    /**
     * The base directory where files get written (excluding package directories)
     * default: "src/test/java"
     *
     * You might want to override if you have tests under "src/test/integration" for example
     *
     * @return snapshot output folder
     */
    default String getOutputDir() {
        return "src/test/java";
    }

    /**
     * Subdirectory to store snapshots in
     *
     * @return name of subdirectory
     */
    default String getSnapshotFolder() {
        return "__snapshots__";
    }

    /**
     * Optional
     * Algorithm to discover what class is under test.  Frameworks often have a better way to access this via frameworks hooks.
     * However, some don't and can use this method to discover it.
     *
     * @return class under test
     */
    Class<?> getTestClass();

    /**
     * Optional
     * Algorithm to discover what class is under test.  Frameworks often have a better way to access this via frameworks hooks.
     * However, some don't and can use this method to discover it.
     *
     * @param testClass class under test
     * @return  method under tests
     */
    Method getTestMethod(Class<?> testClass);

    /**
     * Optional
     *
     * @return snapshots should be updated automatically without verification
     */
    default Optional<String> updateSnapshot() {
        String value = System.getProperty(JVM_UPDATE_SNAPSHOTS_PARAMETER);
        if ( value != null) {
            return Optional.of(value);
        } else {
            return Optional.empty();
        }
    }

    /**
     * Optional
     * Override to supply your own custom serialization function
     *
     * @return custom serialization function
     */
    default SnapshotSerializer getSerializer() {
        return new ToStringSnapshotSerializer();
    }

}
