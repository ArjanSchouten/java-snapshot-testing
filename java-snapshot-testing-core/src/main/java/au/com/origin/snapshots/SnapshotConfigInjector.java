package au.com.origin.snapshots;

public interface SnapshotConfigInjector {
    SnapshotConfig getSnapshotConfig();
}
