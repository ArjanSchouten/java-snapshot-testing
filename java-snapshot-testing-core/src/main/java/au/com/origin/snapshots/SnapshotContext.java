package au.com.origin.snapshots;

import lombok.Builder;
import lombok.Data;
import lombok.RequiredArgsConstructor;

import java.lang.reflect.Method;
import java.nio.file.Path;

@Builder
@Data
@RequiredArgsConstructor
public class SnapshotContext {
    private final Class<?> testClass;
    private final Method testMethod;
    private final String snapshotFilePath;
    private final String snapshotName;
    private final String existingSnapshot;
    private final String incomingSnapshot;
}
