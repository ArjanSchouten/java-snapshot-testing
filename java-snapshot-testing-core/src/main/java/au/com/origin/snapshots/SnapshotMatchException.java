package au.com.origin.snapshots;

public class SnapshotMatchException extends RuntimeException {

    public SnapshotMatchException(String message) {
        super(message);
    }

    public SnapshotMatchException(String message, Throwable cause) {
        super(message, cause);
    }
}
