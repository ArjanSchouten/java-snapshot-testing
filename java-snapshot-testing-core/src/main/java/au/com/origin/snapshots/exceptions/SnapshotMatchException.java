package au.com.origin.snapshots.exceptions;

public class SnapshotMatchException extends RuntimeException {

    public SnapshotMatchException(String message) {
        super(message);
    }

    public SnapshotMatchException(String message, Throwable cause) {
        super(message, cause);
    }
}
