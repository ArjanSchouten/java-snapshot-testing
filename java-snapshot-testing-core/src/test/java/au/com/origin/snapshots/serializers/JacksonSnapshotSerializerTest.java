package au.com.origin.snapshots.serializers;

import au.com.origin.snapshots.SnapshotConfig;
import au.com.origin.snapshots.SnapshotMatchException;
import au.com.origin.snapshots.config.BaseSnapshotConfig;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZonedDateTime;
import java.util.*;

import static au.com.origin.snapshots.SnapshotMatcher.*;

public class JacksonSnapshotSerializerTest {

    private static final SnapshotConfig DEFAULT_CONFIG = new BaseSnapshotConfig() {
        @Override
        public SnapshotSerializer getSerializer() {
            return new JacksonSnapshotSerializer();
        }
    };

    @Test
    public void shouldSerializeDifferentTypesDeterministically() {
        start(DEFAULT_CONFIG);
        expect(new TypeDummy()).toMatchSnapshot();
        validateSnapshots();
    }

    private static enum AnEnum {
        F, A, D, E, G, B, C;
    }

    private final class TypeDummy {
        private final Void aNull = null;
        private final Object anObject = new Object();
        private final byte aByte = "A".getBytes()[0];
        private final short aShort = 32767;
        private final int anInt = 2147483647;
        private final long aLong = 9223372036854775807L;
        private final float aFloat = 0.1234567F;
        private final double aDouble = 1.123456789123456D;
        private final boolean aBoolean = true;
        private final char aChar = 'A';
        private final String string = "Hello World";
        private final Date date = Date.from(Instant.parse("2020-10-19T22:21:07.103Z"));
        private final LocalDate localDate = LocalDate.parse("2020-10-19");
        private final LocalDateTime localDateTime = LocalDateTime.parse("2020-10-19T22:21:07.103");
        private final ZonedDateTime zonedDateTime = ZonedDateTime.parse("2020-10-19T22:21:07.103+10:00[Australia/Melbourne]");
        private final AnEnum anEnum = AnEnum.A;
        private final String[] stringArray = {
            "f",
            "a",
            "d",
            "e",
            "g",
            "b",
            "c"
        };
        private final Object[] anEnumArray = Arrays.stream(AnEnum.values()).toArray();
        private final List<String> arrayList = new ArrayList<String>() {{
            add("f");
            add("a");
            add("d");
            add("e");
            add("g");
            add("b");
            add("c");
        }};
        private final Set<String> hashSet = new HashSet<String>() {{
            add("f");
            add("a");
            add("d");
            add("e");
            add("g");
            add("b");
            add("c");
        }};
        private final Set<String> treeSet = new TreeSet<String>() {{
            add("f");
            add("a");
            add("d");
            add("e");
            add("g");
            add("b");
            add("c");
        }};
        private final Map<String, Integer> hashMap = new HashMap<String, Integer>() {{
            put("f", 6);
            put("a", 1);
            put("d", 4);
            put("e", 5);
            put("g", 7);
            put("b", 2);
            put("c", 3);
        }};
        private final Optional<String> presentOptional = Optional.of("Hello World");
        private final Optional<String> emptyOptional = Optional.empty();
    }
}
