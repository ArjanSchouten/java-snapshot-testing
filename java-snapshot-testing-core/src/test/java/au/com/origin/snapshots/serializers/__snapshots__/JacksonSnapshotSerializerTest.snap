au.com.origin.snapshots.serializers.JacksonSnapshotSerializerTest.shouldSerializeDifferentTypesDeterministically=[
  {
    "anObject": { },
    "aByte": 65,
    "aShort": 32767,
    "anInt": 2147483647,
    "aLong": 9223372036854775807,
    "aFloat": 0.1234567,
    "aDouble": 1.123456789123456,
    "aBoolean": true,
    "aChar": "A",
    "string": "Hello World",
    "date": "2020-10-19T22:21:07.103+00:00",
    "localDate": "2020-10-19",
    "localDateTime": "2020-10-19T22:21:07.103",
    "zonedDateTime": "2020-10-19T22:21:07.103+11:00[Australia/Melbourne]",
    "anEnum": "A",
    "stringArray": [
      "f",
      "a",
      "d",
      "e",
      "g",
      "b",
      "c"
    ],
    "anEnumArray": [
      "F",
      "A",
      "D",
      "E",
      "G",
      "B",
      "C"
    ],
    "arrayList": [
      "f",
      "a",
      "d",
      "e",
      "g",
      "b",
      "c"
    ],
    "hashSet": [
      "a",
      "b",
      "c",
      "d",
      "e",
      "f",
      "g"
    ],
    "treeSet": [
      "a",
      "b",
      "c",
      "d",
      "e",
      "f",
      "g"
    ],
    "hashMap": {
      "a": 1,
      "b": 2,
      "c": 3,
      "d": 4,
      "e": 5,
      "f": 6,
      "g": 7
    },
    "presentOptional": "Hello World",
    "emptyOptional": null
  }
]